### PR TITLE
[Block: Group] Add transform to row variation

### PR DESCRIPTION
**Pull request by carolinan**
## Description

Makes it possible to transform a group into a row variation and vice versa.
For https://github.com/WordPress/gutenberg/issues/35355

## How has this been tested?

1. Add a group block and transform it into a row.
2. Transform it back to a group.

## Screenshots 
Block settings sidebar transformation modal

## Types of changes

Enhancement

## Checklist:
- [x] My code is tested.
- [x] My code follows the WordPress code style. 
- [ ] My code follows the accessibility standards. 
- [ ] I've tested my changes with keyboard and screen readers. 
- [ ] My code has proper inline documentation. 
- [ ] I've included developer documentation if appropriate. 
- [ ] I've updated all React Native files affected by any refactorings/renamings in this PR (please manually search all `*.native.js` files for terms that need renaming or removal).

### CODE REVIEW

File: packages/block-library/src/group/variations.js
Comment (Mamaduka) [inline packages/block-library/src/group/variations.js]:
```suggestion
		scope: [ 'transform' ],
```

To avoid duplicated Group block in the inserter:
CleanShot 2022-01-04 at 18 42 12

Comment (gziolo) [reply packages/block-library/src/group/variations.js]: We could also set this variation as a default one with `isDefault` set to `true` so it doesn't show up twice in the inserter. However, it might have some other consequences because this variation overrides the title and description.

**Comment by carolinan** [comment]
@Mamaduka @aristath Can someone please help me rebase this, so that it can pass the remaning test? I always mess it up and ping 100 contributors.

File: packages/block-library/src/group/variations.js
Comment (youknowriad) [inline packages/block-library/src/group/variations.js:15]: When we insert a group block, it's not checked because an undefined layout is also a "group" block. So we should add `! blockAttributes.layout || blockAttributes.layout?.type === 'default'` here

**Review by youknowriad — COMMENTED** [review]
This works great for me.

It made me wonder whether the variation switcher should be just merged with the regular block switcher (the UI to transform to other blocks) instead of being its own thing.

**Comment by Mamaduka** [comment]
> It made me wonder whether the variation switcher should be just merged with the regular block switcher (the UI to transform to other blocks) instead of being its own thing.

@ndiego had similar questions a few ago, this is the explanation I could find in GitHub archives - https://github.com/WordPress/gutenberg/issues/25231#issuecomment-707909261

**Comment by jasmussen** [comment]
> It made me wonder whether the variation switcher should be just merged with the regular block switcher (the UI to transform to other blocks) instead of being its own thing.

Yep that's been brought up a couple of times, the chief argument being that the difference between the two transforms is a bit hard to understand. The details are fuzzy, but I vaguely recall the challenge being related to getting the mover control for child blocks to switch orientation with the direction, leading this to needing its own separate interface. But if we can combine the two, it makes sense to me, as transforms are so powerful.

Transforms in general need some design explorations, it can become an exceptionally powerful interface and one that you intuitively want to check for each block, as it can contain a multitude of apperance related properties. I'm not sure whether such a design exploration needs to happen before the two interfaces can be merged, but an exploration should definitely take it into account.

As a provocation (and not a strong opinion at all), I want to ask a different question: what if Group was a single block with the Layout orientation switcher? This one:
Screenshot 2022-02-07 at 11 27 00

The question is: is the direction of the content a _transform_ or a _property_? I understand it is in the technical sense, since Group isn't a flex container. But conceptually it's worth thinking about why the interface is different.

**Comment by youknowriad** [comment]
> As a provocation (and not a strong opinion at all), I want to ask a different question: what if Group was a single block with the Layout orientation switcher?

I think that's totally possible, for me the main "pro" of having a different variation is surfacing a "row" in the inserter. Once inserter it matters less whether you should switch on the sidebar or using a variation transform.

**Comment by jasmussen** [comment]
> surfacing a "row" in the inserter

Absolutely. It's also a good term to use in conversation around site building.

Assuming it's possible, it would be good for any design exercises to consider whether we then need both the direction switcher, and the transforms interface between the two, at least in terms of managing prominence.

**Comment by youknowriad** [comment]
@jasmussen I also wanted to clarify that switching between "flex" and "default" layout is not a "direction" switching even if it looks like that. It has more impact than that:

- For "flex" layout, you can use justification
- For "default" layout, you can define "widths" (wide, full), so it also enables these alignments and authorizes "floats" for the child blocks.

For this reason and even if we keep a layout switcher in the block inspector directly, I'm not sure it should be advertised as "direction".

**Comment by jasmussen** [comment]
> For this reason and even if we keep a layout switcher in the block inspector directly, I'm not sure it should be advertised as "direction".

Not suggesting the following, but just in terms of understanding: 

- The group is just a div
- The row is a flexbox, and conceptually if you changed it to vertical, we could call it "Column" :)

**Comment by Mamaduka** [comment]
> The group is just a div

Not anymore 😄 
CleanShot 2022-02-07 at 14 45 38

**Comment by ndiego** [comment]
In some ways, I kinda wish the Row block was it's own block and not a variation of the Group block. It is actually quite different and then a normal block transformation could be done between the two. For example, the "Inherit default layout" selector exists on Row blocks, but shouldn't since it's a flex container. Also, when that setting is enabled it turns the block back into a normal Group block. I think the ship has sailed on this, but having the Row block be its own block would have probably been the easiest scenario.

**Comment by youknowriad** [comment]
@jasmussen Indeed we do have "orientation" support in the "flex" layout directly. So we have:

 - div (layout default aka group)
 - row (layout flex + orientation horizontal)
 - column? (layout flex + orientation vertical): this one exists technically but is hidden, I believe that if we surface it, there will be some confusion with "group" because they are both "vertical" but one has different options than the other.

**Comment by jasmussen** [comment]
> column? (layout flex + orientation vertical): this one exists technically but is hidden, I believe that if we surface it, there will be some confusion with "group" because they are both "vertical" but one has different options than the other.

Right, so "Row" could have the orientation switcher as a way to get to the "Column", perhaps? Just thinking out loud at this point, and it'd generally seem best to abstract the terminology away since they're all just containers with properties.

**Comment by youknowriad** [comment]
@jasmussen yes, the orientation switcher is there but hidden :P

**Comment by youknowriad** [comment]
To avoid blocking this PR which is indeed a nice feature. Can we do this:

 - Address this point and land it https://github.com/WordPress/gutenberg/pull/36202#discussion_r800494318
 - Consider separately whether to add the Column block and allow the "orientation" switcher in the row and column variations. (I think this is probably more useful when we introduce a hight/width control)
 - Consider separately whether we should merge all kind of transforms in the same menu (move variation transforms to the block switcher)

File: packages/block-library/src/group/variations.js
Comment (ntsekouras) [inline packages/block-library/src/group/variations.js:12]: Should we also add here the `inserter` scope and set it as default (`isDefault:true`)?

**Review by youknowriad — APPROVED** [review]
LGTM 👍

**Review by aristath — APPROVED** [review]

**Pull request merged**
